networkx/algorithms/community/louvain.py

```diff
@@ -0,0 +1,289 @@
+
+    Input `m` is the size of the graph `G`.
+    Input `partition` is a valid partition of the graph `G` given as a list of sets of
+    nodes.
+    """
+    node2com = {u: i for i, u in enumerate(G.nodes())}
+    inner_partition = [{u} for u in G.nodes()]
+    if is_directed:
+        in_degrees = dict(G.in_degree(weight="weight"))
+        out_degrees = dict(G.out_degree(weight="weight"))
+        Stot_in = [deg for deg in in_degrees.values()]
+        Stot_out = [deg for deg in out_degrees.values()]
+    else:
+        degrees = dict(G.degree(weight="weight"))
+        Stot = [deg for deg in degrees.values()]
+    nbrs = {u: {v: data["weight"] for v, data in G[u].items() if v != u} for u in G}
+    rand_nodes = list(G.nodes)
+    seed.shuffle(rand_nodes)
+    nb_moves = 1
+    improvement = False
+    while nb_moves > 0:
+        nb_moves = 0
+        for u in rand_nodes:
+            best_mod = 0
+            best_com = node2com[u]
+            weights2com = _neighbor_weights(nbrs[u], node2com)
+            if is_directed:
+                in_degree = in_degrees[u]
+                out_degree = out_degrees[u]
+                Stot_in[best_com] -= in_degree
+                Stot_out[best_com] -= out_degree
+            else:
+                degree = degrees[u]
+                Stot[best_com] -= degree
+            for nbr_com, wt in weights2com.items():
+                if is_directed:
+                    gain = (
+                        wt
+                        - resolution
+                        * (
+                            out_degree * Stot_in[nbr_com]
+                            + in_degree * Stot_out[nbr_com]
+                        )
+                        / m
+                    )
+                else:
+                    gain = 2 * wt - resolution * (Stot[nbr_com] * degree) / m
+                if gain > best_mod:
+                    best_mod = gain
+                    best_com = nbr_com
+            if is_directed:
+                Stot_in[best_com] += in_degree
+                Stot_out[best_com] += out_degree
+            else:
+                Stot[best_com] += degree
+            if best_com != node2com[u]:
+                com = G.nodes[u].get("nodes", {u})
+                partition[node2com[u]].difference_update(com)
+                inner_partition[node2com[u]].remove(u)
+                partition[best_com].update(com)
+                inner_partition[best_com].add(u)
+                improvement = True
+                nb_moves += 1
+                node2com[u] = best_com
+    partition = list(filter(len, partition))
+    inner_partition = list(filter(len, inner_partition))
+    return partition, inner_partition, improvement
+
+
+def _neighbor_weights(nbrs, node2com):
+    """Calculate weights between node and its neighbor communities.
+
+    Input `nbrs` should be a dict of the node's neighbors.
+
+    `node2com` is a dict with nodes as keys and community index as values.
```

Comment:
Formalize the parameters docs for this function.  Say that 'nbrs' is keyed by neighbor to the edge weight as value.